File: day18/day18.py
#!/usr/bin/env python3

import string

class Instruction(object):
    def __init__(self, txt):
        lst = txt.split(" ")
        self.op = lst[0]
        if lst[0] in ("snd", "rcv"):
            self.arg1 = lst[1]
            self.arg2 = None
        elif lst[0] in ("set", "add", "mul", "mod", "jgz"):
            self.arg1 = lst[1]
            self.arg2 = lst[2]

class Duet(object):
    def __init__(self, p = None):
        self.naive = p is None
        self.other = None
        self.send_count = 0
        self.reset(p)

    def reset(self, p):
        self.register = {v: 0 for v in string.ascii_lowercase}
        if not p is None:
            self.register["p"] = p
        self.queue = []
        self.ip = 0
        self.freq = None
        self.exit = False

    def connect(self, other):
        self.other = other

    def resolve(self, value):
        v = self.register.get(value)
        if not v is None:
            return v
        else:
            return int(value)

    def send(self, x):
        self.send_count += 1
        if self.other:
            self.other.queue.append(x)
            self.other.exit = False

    def receive(self):
        if self.queue:
            return self.queue.pop(0)
        self.exit = True

    def play(self, x):
        self.freq = x

    def recover(self):
        self.exit = True

    def execute(self, program):
        while 0 <= self.ip < len(program) and not self.exit:
            i = program[self.ip]
            if i.op == "snd":
                if self.naive:
                    self.play(self.resolve(i.arg1))
                else:
                    self.send(self.resolve(i.arg1))
                self.ip += 1
            elif i.op == "set":
                self.register[i.arg1] = self.resolve(i.arg2)
                self.ip += 1
            elif i.op == "add":
                self.register[i.arg1] += self.resolve(i.arg2)
                self.ip += 1
            elif i.op == "mul":
                self.register[i.arg1] *= self.resolve(i.arg2)
                self.ip += 1
            elif i.op == "mod":
                self.register[i.arg1] %= self.resolve(i.arg2)
                self.ip += 1
            elif i.op == "rcv":
                if self.naive:
                    if self.resolve(i.arg1):
                        self.recover()
                    self.ip += 1
                else:
                    value = self.receive()
                    if not value is None:
                        self.register[i.arg1] = value
                        self.ip += 1
            elif i.op == "jgz":
                if self.resolve(i.arg1) > 0:
                    self.ip += self.resolve(i.arg2)
                else:
                    self.ip += 1

def part1(program):
    d = Duet()
    d.execute(program)
    return d.freq

def part2(program):
    a = Duet(0)
    b = Duet(1)
    a.connect(b)
    b.connect(a)
    while not a.exit or not b.exit:
        a.execute(program)
        b.execute(program)
    return b.send_count

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    program = [Instruction(v) for v in txt.splitlines()]
    print("Part1:", part1(program))
    print("Part2:", part2(program))

File: day18/day18.cpp
#include <fstream>
#include <iostream>
#include <deque>
#include <vector>

#include <fmt/format.h>

using namespace std;

struct Operand
{
	enum { REG, LIT } type;
	int64_t value;
};

istream& operator>>(istream& in, Operand& o)
{
	string tmp;
	in >> tmp;
	if (!in)
	{
		return in;
	}
	if ('a' <= tmp[0] && tmp[0] <= 'z')
	{
		o.type = Operand::REG;
		o.value = tmp[0]-'a';
	}
	else
	{
		o.type = Operand::LIT;
		o.value = stoi(tmp);
	}
	return in;
}

struct Instruction
{
	enum { SND, SET, ADD, MUL, MOD, RCV, JGZ } op;
	Operand arg1;
	Operand arg2;
};

istream& operator>>(istream& in, Instruction& i)
{
	string tmp;
	in >> tmp;
	if (!in)
	{
		return in;
	}
	if (tmp == "snd") i.op = Instruction::SND;
	else if (tmp == "set") i.op = Instruction::SET;
	else if (tmp == "add") i.op = Instruction::ADD;
	else if (tmp == "mul") i.op = Instruction::MUL;
	else if (tmp == "mod") i.op = Instruction::MOD;
	else if (tmp == "rcv") i.op = Instruction::RCV;
	else if (tmp == "jgz") i.op = Instruction::JGZ;
	else
	{
		in.setstate(ios::failbit);
		return in;
	}
	switch (i.op)
	{
	case Instruction::SND:
	case Instruction::RCV:
		in >> i.arg1;
		break;

	default:
		in >> i.arg1;
		in >> i.arg2;
		break;
	}

	return in;
}

struct Duet
{
	int64_t regs['z'-'a'+1];
	int64_t ip;
	int64_t freq;
	deque<int64_t> queue;
	struct Duet *other;
	bool naive;
	bool exit;
	size_t send_count;

	Duet(int64_t program_id = -1)
		: regs{},
		  ip(0),
		  freq(0),
		  other(nullptr),
		  naive(program_id==-1),
		  exit(false),
		  send_count(0)
	{
		if (program_id>=0)
		{
			regs['p'-'a'] = program_id;
		}
	}

	void connect(Duet& o)
	{
		other = &o;
	}

	int64_t resolve(const Operand& o)
	{
		return (o.type == Operand::REG) ? regs[o.value] : o.value;
	}

	void play(int64_t value)
	{
		freq = value;
	}

	void send(int64_t value)
	{
		send_count++;
		if (other)
		{
			other->queue.push_back(value);
			other->exit = false;
		}
	}

	bool receive(int64_t& value)
	{
		if (queue.empty())
		{
			return false;
		}
		value = queue.front();
		queue.pop_front();
		return true;
	}

	void execute(const vector<Instruction>& program)
	{
		while (!exit && 0 <= ip && (size_t)ip < program.size())
		{
			auto &i = program[ip];
			switch (i.op)
			{
			case Instruction::SND:
				naive ? play(resolve(i.arg1)) : send(resolve(i.arg1));
				ip++;
				break;

			case Instruction::SET:
				regs[i.arg1.value] = resolve(i.arg2);
				ip++;
				break;

			case Instruction::ADD:
				regs[i.arg1.value] += resolve(i.arg2);
				ip++;
				break;

			case Instruction::MUL:
				regs[i.arg1.value] *= resolve(i.arg2);
				ip++;
				break;

			case Instruction::MOD:
				regs[i.arg1.value] %= resolve(i.arg2);
				ip++;
				break;

			case Instruction::RCV:
				if (naive)
				{
					if (resolve(i.arg1))
					{
						exit = true;
					}
					ip++;
				}
				else
				{
					if (!receive(regs[i.arg1.value]))
					{
						exit = true;
					}
					else
					{
						ip++;
					}
				}
				break;

			case Instruction::JGZ:
				if (resolve(i.arg1) > 0)
				{
					ip += resolve(i.arg2);
				}
				else
				{
					ip++;
				}
				break;
			}
		}
	}
};

static size_t part1(const vector<Instruction>& program)
{
	Duet d;
	d.execute(program);
	return d.freq;
}

static size_t part2(const vector<Instruction>& program)
{
	Duet a(0), b(1);
	a.connect(b);
	b.connect(a);
	while (!a.exit || !b.exit)
	{
		a.execute(program);
		b.execute(program);
	}
	return b.send_count;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	vector<Instruction> program;
	Instruction instr;
	while (input>>instr)
	{
		program.emplace_back(move(instr));
	}
	input.close();

	cout << fmt::format("Part1: {}", part1(program)) << endl
	     << fmt::format("Part2: {}", part2(program)) << endl;
	return 0;
}
